<a href="https://colab.research.google.com/github/Deok-Hun/Aiffel_Kun/blob/master/%EC%99%80%EC%9D%B8%EC%9D%84_%EB%B6%84%EB%A5%98%ED%95%B4_%EB%B4%85%EC%8B%9C%EB%8B%A4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import drive
drive.mount('/content/drive')
drive.mount("/content/drive", force_remount=True)

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
Mounted at /content/drive


# 와인을 분류 데이터 불러오기

In [2]:
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

In [3]:
wine=load_wine()
print(dir(wine))

['DESCR', 'data', 'feature_names', 'frame', 'target', 'target_names']


In [4]:
wine_data=wine.data
wine_label=wine.target

print(wine_data.shape)


(178, 13)


In [5]:
wine.target_names

array(['class_0', 'class_1', 'class_2'], dtype='<U7')

In [6]:
print(wine.DESCR)

.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

    :Number of Instances: 178 (50 in each of three classes)
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline

    - class:
            - class_0
            - class_1
            - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= =====
                                   Min   Max   Mean     SD
    ============================= ==== ===== ======= =====
    Alcohol:                      11.0  14.8    13.0   0.8
    Malic Acid:                   0.74  5.80    2.34  1.12
    Ash:                          1.36  3.23    2.36  0.27
    Alcalinity of Ash:            1

In [7]:
wine.feature_names

['alcohol',
 'malic_acid',
 'ash',
 'alcalinity_of_ash',
 'magnesium',
 'total_phenols',
 'flavanoids',
 'nonflavanoid_phenols',
 'proanthocyanins',
 'color_intensity',
 'hue',
 'od280/od315_of_diluted_wines',
 'proline']

In [8]:
import pandas as pd
wine_df=pd.DataFrame(data=wine_data,columns=wine.feature_names)
wine_df

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0      14.23        1.71  2.43               15.6      127.0           2.80   
1      13.20        1.78  2.14               11.2      100.0           2.65   
2      13.16        2.36  2.67               18.6      101.0           2.80   
3      14.37        1.95  2.50               16.8      113.0           3.85   
4      13.24        2.59  2.87               21.0      118.0           2.80   
..       ...         ...   ...                ...        ...            ...   
173    13.71        5.65  2.45               20.5       95.0           1.68   
174    13.40        3.91  2.48               23.0      102.0           1.80   
175    13.27        4.28  2.26               20.0      120.0           1.59   
176    13.17        2.59  2.37               20.0      120.0           1.65   
177    14.13        4.10  2.74               24.5       96.0           2.05   

     flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0          3.06                  0.28             2.29             5.64  1.04   
1          2.76                  0.26             1.28             4.38  1.05   
2          3.24                  0.30             2.81             5.68  1.03   
3          3.49                  0.24             2.18             7.80  0.86   
4          2.69                  0.39             1.82             4.32  1.04   
..          ...                   ...              ...              ...   ...   
173        0.61                  0.52             1.06             7.70  0.64   
174        0.75                  0.43             1.41             7.30  0.70   
175        0.69                  0.43             1.35            10.20  0.59   
176        0.68                  0.53             1.46             9.30  0.60   
177        0.76                  0.56             1.35             9.20  0.61   

     od280/od315_of_diluted_wines  proline  
0                            3.92   1065.0  
1                            3.40   1050.0  
2                            3.17   1185.0  
3                            3.45   1480.0  
4                            2.93    735.0  
..                            ...      ...  
173                          1.74    740.0  
174                          1.56    750.0  
175                          1.56    835.0  
176                          1.62    840.0  
177                          1.60    560.0  

[178 rows x 13 columns]

In [9]:
len(wine_df)-wine_df.count()

alcohol                         0
malic_acid                      0
ash                             0
alcalinity_of_ash               0
magnesium                       0
total_phenols                   0
flavanoids                      0
nonflavanoid_phenols            0
proanthocyanins                 0
color_intensity                 0
hue                             0
od280/od315_of_diluted_wines    0
proline                         0
dtype: int64

In [10]:
wine_df["label"]=wine.target
wine_df

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0      14.23        1.71  2.43               15.6      127.0           2.80   
1      13.20        1.78  2.14               11.2      100.0           2.65   
2      13.16        2.36  2.67               18.6      101.0           2.80   
3      14.37        1.95  2.50               16.8      113.0           3.85   
4      13.24        2.59  2.87               21.0      118.0           2.80   
..       ...         ...   ...                ...        ...            ...   
173    13.71        5.65  2.45               20.5       95.0           1.68   
174    13.40        3.91  2.48               23.0      102.0           1.80   
175    13.27        4.28  2.26               20.0      120.0           1.59   
176    13.17        2.59  2.37               20.0      120.0           1.65   
177    14.13        4.10  2.74               24.5       96.0           2.05   

     flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0          3.06                  0.28             2.29             5.64  1.04   
1          2.76                  0.26             1.28             4.38  1.05   
2          3.24                  0.30             2.81             5.68  1.03   
3          3.49                  0.24             2.18             7.80  0.86   
4          2.69                  0.39             1.82             4.32  1.04   
..          ...                   ...              ...              ...   ...   
173        0.61                  0.52             1.06             7.70  0.64   
174        0.75                  0.43             1.41             7.30  0.70   
175        0.69                  0.43             1.35            10.20  0.59   
176        0.68                  0.53             1.46             9.30  0.60   
177        0.76                  0.56             1.35             9.20  0.61   

     od280/od315_of_diluted_wines  proline  label  
0                            3.92   1065.0      0  
1                            3.40   1050.0      0  
2                            3.17   1185.0      0  
3                            3.45   1480.0      0  
4                            2.93    735.0      0  
..                            ...      ...    ...  
173                          1.74    740.0      2  
174                          1.56    750.0      2  
175                          1.56    835.0      2  
176                          1.62    840.0      2  
177                          1.60    560.0      2  

[178 rows x 14 columns]

In [11]:
x_train,x_test,y_train,y_test=train_test_split(wine_data,wine_label,test_size=0.2,random_state=16)

In [12]:
x_train.shape, y_train.shape

((142, 13), (142,))

In [13]:
x_test.shape, y_test.shape

((36, 13), (36,))

# Decision Tree(의사결정나무)

In [14]:
from sklearn.tree import DecisionTreeClassifier

decision_tree=DecisionTreeClassifier(random_state=16)
print(decision_tree._estimator_type)

classifier


In [15]:
decision_tree.fit(x_train,y_train)
y_pred=decision_tree.predict(x_test)
y_pred

array([2, 2, 0, 0, 2, 1, 0, 2, 2, 2, 0, 2, 0, 0, 1, 0, 1, 1, 1, 1, 0, 2,
       0, 1, 2, 2, 2, 1, 2, 1, 2, 0, 2, 0, 2, 1])

In [16]:
from sklearn.metrics import accuracy_score
accuracy=accuracy_score(y_test,y_pred)
accuracy

0.8333333333333334

In [17]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.91      1.00      0.95        10
           1       0.70      0.70      0.70        10
           2       0.87      0.81      0.84        16

    accuracy                           0.83        36
   macro avg       0.83      0.84      0.83        36
weighted avg       0.83      0.83      0.83        36



# Random Forest

In [18]:
from sklearn.ensemble import RandomForestClassifier

x_train, x_test, y_train, y_test = train_test_split(wine_data, wine_label, test_size=0.2, random_state=16)

random_forest=RandomForestClassifier(random_state=16)
random_forest.fit(x_train,y_train)
y_pred=random_forest.predict(x_test)

print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        10
           1       1.00      0.90      0.95        10
           2       0.94      1.00      0.97        16

    accuracy                           0.97        36
   macro avg       0.98      0.97      0.97        36
weighted avg       0.97      0.97      0.97        36



# Support Vector Machine(SVM)

In [19]:
from sklearn import svm
svm_model=svm.SVC()

svm_model.fit(x_train,y_train)
y_pred=svm_model.predict(x_test)

print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.82      0.90      0.86        10
           1       0.36      0.90      0.51        10
           2       0.00      0.00      0.00        16

    accuracy                           0.50        36
   macro avg       0.39      0.60      0.46        36
weighted avg       0.33      0.50      0.38        36



/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


# Stochastic Gradient Descent Classifier (SGD Classifier)

In [20]:
from sklearn.linear_model import SGDClassifier
sgd_model=SGDClassifier()

sgd_model.fit(x_train,y_train)
y_pred=sgd_model.predict(x_test)

print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       1.00      0.70      0.82        10
           1       0.43      0.90      0.58        10
           2       0.62      0.31      0.42        16

    accuracy                           0.58        36
   macro avg       0.68      0.64      0.61        36
weighted avg       0.67      0.58      0.58        36



# Logistic Regression

In [22]:
from sklearn.linear_model import LogisticRegression
logistic_model=LogisticRegression()

logistic_model.fit(x_train,y_train)
y_pred=logistic_model.predict(x_test)

print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        10
           1       0.90      0.90      0.90        10
           2       0.94      0.94      0.94        16

    accuracy                           0.94        36
   macro avg       0.95      0.95      0.95        36
weighted avg       0.94      0.94      0.94        36



/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG,


# 오차 행렬

In [23]:
from sklearn.metrics import confusion_matrix

confusion_matrix(y_test,y_pred)

array([[10,  0,  0],
       [ 0,  9,  1],
       [ 0,  1, 15]])

# 결론

* 학습된 모델들의 테스트데이터 예측결과는 다음과 같다.
  - Decision Tree: 83%
  - Support Vector Machine : 50%
    - UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior._warn_prf(average, modifier, msg_start, len(result): 이는 오류 코드가 아닌 y_pred 안에 모든 label들이 포함되지 않았다는 경고이다.
 즉, 몇몇 label 들이 분류자가 y_test에 예측 할 수 없는 label들이 있다는 뜻이 도출 되었다.
  - tochastic Gradient Descent Classifier : 58%
  - Logistic Regression : 94%

* 와인 분류 데이터의 Feature 수는 13개로 많은 특성을 보유하고 있다.
따라서, 이러한 경우 의사결정나무 보다는 Random Forest 가 더 정확성이 높은
결과를 분류를 할 수 있다.

* 와인의 분류는 제품의 원산지를 확인하는 경우를 비교해 본다.
예를 들어 이것이 중국산 마늘인지 국산 마늘인지를 예측한다고 가정했을 때,
한국산 마늘을 중국산이라고 예측하는 경우보다 중국산 마늘을 한국산 마늘이라고 예측하는 경우가
더 많은 영향을 미친다. 따라서, 한국산이 아닌데 한국산이라고 (FP) 하는 경우인
Precision 이 더 중요한 평가지표이다. 
